skbio/diversity/beta/_unifrac.py

```diff
@@ -189,44 +174,341 @@ def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+
+    return (branch_lengths * abs(u_ - v_)).sum()
+
+
+def _weighted_unifrac_normalized(u_node_counts, v_node_counts, u_total_count,
+                                 v_total_count, tree, tree_index):
+    """
+    Parameters
+    ----------
+    u_node_counts, v_node_counts : np.array
+         Vectors indicating presense (value greater than zero) and absense
+         (value equal to zero) of nodes in two samples, `u` and `v`. Order is
+         assumed to be the same as in `branch_lengths`.
+    u_total_count, v_total_counts : int
+         The sum of ``u_node_counts`` and ``v_node_counts`` vectors,
+         respectively. This could be computed internally, but since this is a
+         private method and the calling function has already generated these
+         values, this saves an iteration over each of these vectors.
+    tree: skbio.TreeNode
+         Tree relating the OTUs.
+    tree_index:
+         Result of calling ``tree.to_array``. This is expensive to compute, so
+         it's provided as a parameter here as the calling function will have
+         already generated it.
+
+    Returns
+    -------
+    float
+        Normalized weighted UniFrac distance between samples.
 
+    Notes
+    -----
+    The count vectors passed here correspond to all nodes in the tree, not
+    just the tips.
+
+    """
+    branch_lengths = tree_index['length']
+    u = _weighted_unifrac(u_node_counts, v_node_counts, u_total_count,
+                          v_total_count, branch_lengths)
+    # get the index positions for tips in counts_array, and determine the
+    # tip distances to the root
+    tip_indices = np.array([n.id for n in tree_index['id_index'].values()
+                            if n.is_tip()])
+    node_to_root_distances = _tip_distances(branch_lengths, tree, tip_indices)
+    u /= _weighted_unifrac_branch_correction(
+        node_to_root_distances, u_node_counts, v_node_counts, u_total_count,
+        v_total_count)
+
+    return u
+
+
+def _unweighted_unifrac_pdist_f(counts, otu_ids, tree):
+    """ Create optimized pdist-compatible unweighted UniFrac function
+
+    Parameters
+    ----------
+    counts : 2D array_like of ints or floats
+        Matrix containing count/abundance data where each row contains counts
+        of observations in a given sample.
+    otu_ids: list, np.array
+        Vector of OTU ids corresponding to tip names in ``tree``. Must be the
+        same length as ``u_counts`` and ``v_counts``. These IDs do not need to
+        be in tip order with respect to the tree.
+    tree: skbio.TreeNode
+        Tree relating the OTUs in otu_ids. The set of tip names in the tree can
+        be a superset of ``otu_ids``, but not a subset.
+
+    Returns
+    -------
+    function
+        Optimized pairwise unweighted UniFrac calculator that can be passed
+        to ``scipy.spatial.distance.pdist``.
+    2D np.array of ints, floats
+        Counts of all nodes in ``tree``.
+    2D np.array of floats
```

Comment:
I don't believe it is necessary. If it is, important to note that `branch_lengths` I believe is a 1D array not 2D.